subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy

```diff
@@ -16,39 +16,42 @@
 
 package org.gradle.internal.reflect
 
-import com.google.common.reflect.TypeToken
+
 import org.gradle.testing.internal.util.Specification
 import spock.lang.Unroll
 
-import java.util.function.BiConsumer
+import java.lang.reflect.Type
 
 class JavaReflectionUtilTest extends Specification {
 
     @Unroll
     def "#type has type variable: #hasTypeVariable"() {
         expect:
-        JavaReflectionUtil.hasTypeVariable(type.getType()) == hasTypeVariable
+        JavaReflectionUtil.hasTypeVariable(type) == hasTypeVariable
+
+        where:
+        testType << testedTypes
+        type = testType.first
+        hasTypeVariable = testType.second
+    }
+
+    @Unroll
+    def "resolving return type #method.genericReturnType does not have type parameters"() {
+        expect:
+        !JavaReflectionUtil.hasTypeVariable(
+            JavaReflectionUtil.resolveMethodReturnType(JavaReflectionUtilTestMethods.InterfaceRealizingTypeParameter, method)
+        )
 
         where:
-        type                                                              | hasTypeVariable
-        new TypeToken<Class<?>>() {}                                      | false
-        new TypeToken<Class<String>>() {}                                 | false
-        new TypeToken<Class<String>[]>() {}                               | false
-        TypeToken.of(String.class)                                        | false
-        TypeToken.of(String[].class)                                      | false
-        new TypeToken<List<BiConsumer<String, Collection<Integer>>>>() {} | false
-        new TypeToken<List<? extends Collection<String>>>() {}            | false
-        new TypeToken<List<? super List<String>>>() {}                    | false
-        new TypeToken<BiConsumer<List<String>, List<String>>>() {}        | false
-        genericReturnType("simpleGenericReturnType")                      | true
-        genericReturnType("encapsulatedTypeVariable")                     | true
-        genericReturnType("arrayTypeWithTypeVariable")                    | true
-        genericReturnType("complexTypeWithTypeVariable")                  | true
-        genericReturnType("anotherComplexTypeWithTypeVariable")           | true
-        genericReturnType("complexTypeWithArrayTypeVariable")             | true
+        method << (JavaReflectionUtilTestMethods.getDeclaredMethods() as List)
```

Comment:
Looks better, though shouldn't we check the actual type to resolve to? You could use an annotation to attach the expected type to the methods or something.

Reply:
If we did that, maybe it would make sense to add a test to check if `<X extends String> X getSomething()` indeed resolves to `String`.

Reply:
IIUC, the generic return type of `<X extends String> X getSomething()` resolves to `<X extends String>`.

Reply:
Alright. Can we check it?

Reply:
Checking now.